tests/python/tenpy/gates/ucj_test.py

```diff
@@ -0,0 +1,101 @@
+# that they have been altered from the originals.
+
+"""Tests for the TeNPy unitary cluster Jastrow gate."""
+
+import numpy as np
+import pytest
+from tenpy.algorithms.tebd import TEBDEngine
+
+import ffsim
+from ffsim.tenpy.gates.ucj import apply_ucj_op_spin_balanced
+from ffsim.tenpy.hamiltonians.molecular_hamiltonian import MolecularHamiltonianMPOModel
+from ffsim.tenpy.util import bitstring_to_mps
+from ffsim.testing import interaction_pairs_spin_balanced
+
+
+@pytest.mark.parametrize(
+    "norb, nelec, n_reps, connectivity",
+    [
+        (4, (2, 2), 1, "square"),
+        (4, (1, 2), 1, "square"),
+        (4, (0, 2), 1, "square"),
+        (4, (0, 0), 1, "square"),
+        (4, (2, 2), 1, "hex"),
+        (4, (1, 2), 1, "hex"),
+        (4, (0, 2), 1, "hex"),
+        (4, (0, 0), 1, "hex"),
+        (4, (2, 2), 1, "heavy-hex"),
+        (4, (1, 2), 1, "heavy-hex"),
+        (4, (0, 2), 1, "heavy-hex"),
+        (4, (0, 0), 1, "heavy-hex"),
+        (4, (2, 2), 2, "square"),
+        (4, (1, 2), 2, "square"),
+        (4, (0, 2), 2, "square"),
+        (4, (0, 0), 2, "square"),
+        (4, (2, 2), 2, "hex"),
+        (4, (1, 2), 2, "hex"),
+        (4, (0, 2), 2, "hex"),
+        (4, (0, 0), 2, "hex"),
+        (4, (2, 2), 2, "heavy-hex"),
+        (4, (1, 2), 2, "heavy-hex"),
+        (4, (0, 2), 2, "heavy-hex"),
+        (4, (0, 0), 2, "heavy-hex"),
+    ],
+)
+def test_apply_ucj_op_spin_balanced(
+    norb: int, nelec: tuple[int, int], n_reps: int, connectivity: str
+):
+    """Test applying a spin-balanced unitary cluster Jastrow gate to an MPS."""
+    rng = np.random.default_rng()
+
+    # generate a random molecular Hamiltonian
+    mol_hamiltonian = ffsim.random.random_molecular_hamiltonian(norb, seed=rng)
+    hamiltonian = ffsim.linear_operator(mol_hamiltonian, norb, nelec)
+
+    # convert molecular Hamiltonian to MPO
+    mol_hamiltonian_mpo_model = MolecularHamiltonianMPOModel.from_molecular_hamiltonian(
+        mol_hamiltonian
+    )
+    mol_hamiltonian_mpo = mol_hamiltonian_mpo_model.H_MPO
+
+    # generate a random LUCJ ansatz
+    lucj_op = ffsim.random.random_ucj_op_spin_balanced(
+        norb=norb,
+        n_reps=n_reps,
+        interaction_pairs=interaction_pairs_spin_balanced(
+            connectivity=connectivity, norb=norb
+        ),
+        with_final_orbital_rotation=True,
+        seed=rng,
+    )
+
+    # generate the corresponding LUCJ circuit statevector
+    hf_state = ffsim.hartree_fock_state(norb, nelec)
+    lucj_state = ffsim.apply_unitary(hf_state, lucj_op, norb, nelec)
+
+    # generate the corresponding LUCJ circuit MPS
+    dim = ffsim.dim(norb, nelec)
+    strings_a, strings_b = ffsim.addresses_to_strings(
+        range(dim),
+        norb=norb,
+        nelec=nelec,
+        bitstring_type=ffsim.BitstringType.STRING,
+        concatenate=False,
+    )
+    wavefunction_mps = bitstring_to_mps((strings_a[0], strings_b[0]))
```

Comment:
```suggestion
    n_alpha, n_beta = nelec
    wavefunction_mps = bitstring_to_mps(((1 << n_alpha) - 1, (1 << n_beta) - 1))
```

Reply:
That is much cleaner 🙂

Reply:
Is there also a compact way of writing `bitstring_to_mps((strings_a[idx], strings_b[idx]))`, where `idx` is a random integer ?

Reply:
More generally, I am wondering about the best API for the function `bitstring_to_mps`. Currently, it takes `tuple[int, int]` and `norb`. However, it could instead take simply `int[str, str]`. I have also noticed that you have a `BitstringType` class with some conversion functions.